### ISSUE TITLE
Publishing from Windows Subsystem Linux (WSL Ubuntu distro) terminal or browser fails to create a draft in Medium

### ISSUE THREAD

**Issue by etzimopoulos** [dexplo/jupyter_to_medium#4]
I'm running Windows 10 and I have been using an Ubuntu Linux distro under the Windows Linux Subsystem.

When I click on "Deploy to Medium" I get to the publish screen without any issues, but then when I click on "Publish" I get the following error:

```
Failed to post to Medium
Returned error message below

"cannot identify image file <_io.BytesIO object at 0x7faefba705f0>"
```

```
When I publish from the terminal I get a more detailed view of the issue:
Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Permission denied
Failed to generate minidump.Traceback (most recent call last):
  File "/home/etzimopoulos/anaconda3/bin/jupyter_to_medium", line 8, in <module>
    sys.exit(main())
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_command_line.py", line 129, in main
    publish(**args)
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_publish_to_medium.py", line 289, in publish
    p.main()
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_publish_to_medium.py", line 216, in main
    self.md, self.image_data_dict = self.create_markdown()
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_publish_to_medium.py", line 114, in create_markdown
    self.nb, self.resources = no_ex_pp.preprocess(self.nb, self.resources)
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/nbconvert/preprocessors/base.py", line 69, in preprocess
    nb.cells[index], resources = self.preprocess_cell(cell, resources, index)
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_preprocesors.py", line 126, in preprocess_cell
    output['data'] = {'image/png': ss_creator(html)}
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_screenshot.py", line 147, in _repr_png_
    return self.run(html)
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_screenshot.py", line 136, in run
    img = self.finalize_image(buffer)
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/jupyter_to_medium/_screenshot.py", line 108, in finalize_image
    img = Image.open(buffer)
  File "/home/etzimopoulos/anaconda3/lib/python3.7/site-packages/PIL/Image.py", line 2896, in open
    "cannot identify image file %r" % (filename if filename else fp)
PIL.UnidentifiedImageError: cannot identify image file <_io.BytesIO object at 0x7f9606bf7b30>
```

It looks like it's having an issue saving down the image using the **Pillow** library and that fails under **"_screenshot.py"**

**Background**
One of the issues I've had, in getting that far with using the library, was the fact that I'm running WSL and chrome browser would not always be accessed correctly by various libraries. e.g. - I wasn't able to launch a browser tab when I was launching my Jupyter Notebook under the WSL ubuntu distro as there was no UI and so I had to point it to the Windows Chrome executable (See details [here](https://stackoverflow.com/a/62275339/11141700))
- I suspect the libraries are using the /usr/bin/chrome path to access chrome but under my configuration it wouldn't be able to launch the final confirmation about the draft medium page. To be honest I'm a little confused at this stage as to how these interplay, i.e. the Linux Chrome Installation (/usr/bin/chrome) vs the windows executable which needs to be used at some point to see the results published.
- To verify this, I installed the other library you've published (dataframe_image) to generate an image from dataframes into a PDF, and again, when I publish the final PDF, I get the following error when using the option "Download-->DataFrame by Image via Browser (PDF)": 
```
500 : Internal Server Error
The error was:

nbconvert failed: Could not connect to chrome server
```

To summarise:
1. I'm either missing some configuration in my setup to interchange between Windows Chrome and WSL Linux Chrome or there must be issue in the libraries that needs fixing/updating to cater for WSL (or a combination).
2. The actual issue of failing to convert the image using screenshot might be a genuine bug as I run the library with both chrome versions and got the same error (unless within the library it's using native linux chrome and that's the reason it's failing anyway).

Appreciate it's a long post but any support and feedback would be great.

Thanks

**Comment by tdpetrou** [comment]
Can you attach your notebook here so I can download it and test it on my end?

**Comment by etzimopoulos** [comment]
Yes of course, thank you. I've played around with many notebooks but this simple one would do whereby I've tried to use some of your demo commands for generating a dataframe as the output of a cell and then highlighting it.

I also created a plot graph to see if this comes through but still no luck.
[TestCellOutput.zip](https://github.com/dexplo/jupyter_to_medium/files/4817403/TestCellOutput.zip)

**Comment by tdpetrou** [comment]
Thanks. Also, can you run the following lines and tell me output?

```
import platform
platform.system()
platform.platform()
```

**Comment by etzimopoulos** [comment]
Here it is: 
'Linux-4.4.0-19041-Microsoft-x86_64-with-debian-bullseye-sid'

I imagine this is going to be added as an option under the Operating systems or so? Something to pick upon in the IF statement? :)

**Comment by tdpetrou** [comment]
Thanks. No, that's `platform.platform`. I'm guessing `platform.system` is `'Linux'`

**Comment by etzimopoulos** [comment]
![image](https://user-images.githubusercontent.com/36522537/85413168-6094ac00-b562-11ea-98e3-1bfbc28cb48b.png)
Yes you're right, apologies, I run both on the same cell previously.

**Comment by tdpetrou** [comment]
No problem. I tested your notebook on my machine and it works.

It's finding chrome and taking screenshots, just reading it as 0 bytes. We need to find the chrome executable on your machine: It looks in one of these places
```
None,
 "/usr/local/sbin",
"/usr/local/bin",
"/usr/sbin",
"/usr/bin",
"/sbin",
"/bin",
"/opt/google/chrome"
```

for one of these commands:

```
"google-chrome", "chrome", "chromium", "chromium-browser", "brave-browser"
```

Can you identify where it is and which command its calling?

**Comment by etzimopoulos** [comment]
So I'm running Ubuntu under Windows Subsystem Linux, which means, there's no browser only the terminal. 

I had a similar issue with launching Jupyter Notebook itself from Ubuntu, and the way I resolved it was pointing to the windows Chrome Executable (See [here](https://stackoverflow.com/a/62275339/11141700) )

The full path is 
`/mnt/c/Program Files (x86)/Google/Chrome/Application/chrome.exe`

which I also used when running the library as an argument, but made no difference:
` jupyter_to_medium TestCellOutput.ipynb --chrome-path='/mnt/c/Program Files (x86)/Google/Chrome/Application/chrome.exe' `

At some point I edited your files locally to add this path to the ones you have above, but didn't work (or maybe I did something wrong).

**Comment by etzimopoulos** [comment]
I realised I forgot to address the other question. I just tested it and it's "google-chrome" but for some reason it's not working. This is what I get when I run it:

`google-chrome
Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Permission denied
Trace/breakpoint trap (core dumped)
(base) etzimopoulos@DESKTOP-HI9P6E4:/mnt/c/Users/etzim/Google Drive/Career/Data Science/GitHub/jupyter-to-medium$ [275:275:0100/000000.539345:ERROR:zygote_linux.cc(653)] write: Broken pipe (32)
[0623/155257.539491:ERROR:nacl_helper_linux.cc(308)] NaCl helper process running without a sandbox!
Most likely you need to configure your SUID sandbox correctly`

So it looks like I need to fix that on my end. I would have expected "chrome" to work given I'm able to launch a browser with Jupyter from Ubuntu, but I'm fairly new to this cross platform functionality.

**Comment by tdpetrou** [comment]
Thanks for all the information. Can you try taking screenshots with chrome headless from the terminal

https://developers.google.com/web/updates/2017/04/headless-chrome#screenshots

Command will look something like this

```
/mnt/c/Program Files (x86)/Google/Chrome/Application/chrome.exe --headless --disable-gpu --screenshot https://www.chromestatus.com/
```

**Comment by etzimopoulos** [comment]
Hi Ted,

I'm getting a little confused here so I'll describe the things I've tried:

1) Running the command you suggested doesn't do anything as it gets stuck running forever. Could this be because it's a windows executable and we're passing linux arguments (e.g. headless, screenshot, etc...)?

2) I tried the same with google-chrome and got this:
` etzimopoulos@DESKTOP-HI9P6E4:~google-chrome --headless --disable-gpu --screenshot https://www.chromestatus.com/
Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Permission denied
Failed to generate minidump.Illegal instruction (core dumped)`

3) If I run google-chrome on my terminal I get the following (gets stuck running):
` etzimopoulos@DESKTOP-HI9P6E4:~$ google-chrome
Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Permission denied
Trace/breakpoint trap (core dumped)
(base) etzimopoulos@DESKTOP-HI9P6E4:~$ [1305:1305:0100/000000.631628:ERROR:zygote_linux.cc(653)] write: Broken pipe (32)
[0624/070546.631832:ERROR:nacl_helper_linux.cc(308)] NaCl helper process running without a sandbox!
Most likely you need to configure your SUID sandbox correctly`

4) I played around a bit and managed to make this work (but temporarilly)
`google-chrome --no-gpu --no-sandbox --disable-setuid-sandbox --headless --screenshot http://www.chromestatus.com`

It created an image along with an error, but unfortunately worked only once. See attached picture here:
![screenshot](https://user-images.githubusercontent.com/36522537/85507900-fa07a080-b5ea-11ea-88b7-9cdd3598e3c2.png)

5) Tried running this again but it now keeps getting stuck without saving the image:
`(base) etzimopoulos@DESKTOP-HI9P6E4:~$ google-chrome --no-gpu --no-sandbox --disable-setuid-sandbox --headless --screenshot http://www.chromestatus.com
[0624/072025.641074:ERROR:udev_watcher.cc(61)] Failed to enable receiving udev events.
`

**Comment by tdpetrou** [comment]
Not really sure either. Can you run the commands again with `sudo`?

**Comment by tdpetrou** [comment]
I just uploaded a new version 0.1.0 to pypi with the option of using matplotlib to create tables instead of google chrome. The tables will look a bit different, but should be pretty close to the original.

Can you try with the new option and report back?

**Comment by etzimopoulos** [comment]
Hi Ted, 

Unfortunately none of these worked. I think it's an issue with my system setup (Windows Subsystem Linux-WSL). 

Let me take this away and I'll come back to you if I have any updates. I'll try to setup a VM to run this on Linux properly to check if its me or the library :)

Thanks again for your support. 

Angelo

**Comment by tdpetrou** [comment]
Did you try the matplotlib option? It has nothing to do with chrome.

**Comment by etzimopoulos** [comment]
It's working!!! Hooray!

Sorry, didn't realise it was an added parameter so previous test was invalid. Here's what I tried:

**1) TestCellOutput.ipynb - via the terminal**

I run the following command: `jupyter_to_medium TestCellOutp - WORKED FINE
ut.ipynb --chrome-path='/mnt/c/Program Files (x86)/Google/Chrome/Application/chrome.exe' --table-conversion='matplotlib'`

Output:
```
findfont: Font family ['Helvetica'] not found. Falling back to DejaVu Sans.
findfont: Font family ['Helvetica'] not found. Falling back to DejaVu Sans.
loading image to medium
loading image to medium
loading image to medium

Successfully posted to Medium!!!
--------------------------------
id                  xxxx
title               TestCellOutput
authorId           xxxx
url                 https://medium.com/@etzimopoulos/xxxx
canonicalUrl
publishStatus       draft
license             all-rights-reserved
licenseUrl          https://medium.com/policy/9db0094a1e0f
tags                []
```


**2) TestCellOutput.ipynb via the UI** (selecting Matplotlib option in the publishing page) -  WORKED FINE


**3) TestJ2m.ipynb via the UI**  - WORKED FINE
This was the original I was trying to publish as an article when I realised I'm getting errors so tested that to make sure it works.

Attached both files below.

[TestJ2m.zip](https://github.com/dexplo/jupyter_to_medium/files/4839059/TestJ2m.zip)

**Comment by etzimopoulos** [comment]
Also, this was the log from running Test2m.ipynb via the UI:

```
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
findfont: Font family ['Helvetica'] not found. Falling back to DejaVu Sans.
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
Fontsize 0.00 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
loading image to medium
loading image to medium
loading image to medium
loading image to medium
loading image to medium
loading image to medium
loading image to medium



Successfully posted to Medium!!!
--------------------------------
id                  f896d38a16ad
title               TestJ2m
```

**Comment by tdpetrou** [comment]
Excellent! Thanks for testing again. I'll try and make the matplotlib conversion better in the future.

As for WSL. I have no idea how to make it work and don't really want to use my time to dig deeper into it as the matplotlib option should be good enough as a replacement.
